File: TypeScript/0704-Binary-Search.js
function search(nums, target) {
    var mid, left = 0, right = nums.length;
    while (left < right) {
        mid = left + ((right - left) >> 1);
        if (nums[mid] > target) {
            right = mid;
        }
        else if (nums[mid] < target) {
            left = mid + 1;
        }
        else {
            return mid;
        }
    }
    return -1;
}
;
var nums = [1, 2, 3, 4, 5, 6];
var target = 3;
var result = search(nums, target);
console.log(result); // 2

File: Python/0704-Binary-Search.py
from typing import List

class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        l = 0
        r = len(nums)
        while l < r:
            mid = (l + r) // 2
            if nums[mid] < target:
                l = mid + 1
            elif nums[mid] > target:
                r = mid
            else:
                return mid
        return -1

class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        l = 0
        r = len(nums) - 1
        while l <= r:
            mid = (l + r) // 2
            if nums[mid] < target:
                l = mid + 1
            elif nums[mid] > target:
                r = mid - 1
            else:
                return mid
        return -1

if __name__ == "__main__":
    nums = [-1, 0, 3, 5, 9, 12]
    ###nums = [5]
    result = Solution().search(nums, 5)
    print(result)
